File: Data Structures & Algorithms/decode-ways/submission-2.py
class Solution:
    def numDecodings(self, s: str) -> int:
        n = len(s)

        def dfs(i):
            if i == len(s):
                return 1
            if s[i] == "0":
                return 0
            
            if i + 2 <= n and int(s[i:i + 2]) <= 26:
                return dfs(i + 2) + dfs(i + 1)
            else:
                return dfs(i + 1)

        return dfs(0) 
        

File: Data Structures & Algorithms/climbing-stairs/submission-2.cpp
class Solution {
public:
    int climbStairs(int n) {
        return dfs(n, 0);
    }
    int dfs(int n, int pos){
        int numClimbs = 0;

        // base case
        if (pos == n) {
            numClimbs++;
        }
        else if (pos > n) {
            return 0;
        }
        else {
            numClimbs += dfs(n, pos + 1);
            numClimbs += dfs(n, pos + 2);
        }
        return numClimbs;

    }
};
// 1+1=2
// 1+2=3
// 2+3=5
// 3+5=8


// public:
//     int climbStairs(int n) {
//         if (n == 1) {
//             return 1;
//         }
//         if (n == 2) {
//             return 2;
//         }
//         int first = 1;
//         int second = 1;
//         for (int i = 2; i < n; i++) {
//             int sum = first + second;
//             first = second;
//             second = sum;
//         }
//         return first + second;
//     }
// };